src/algebra/ne_zero.lean

```diff
@@ -0,0 +1,66 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import algebra.algebra.basic
+import algebra.char_p.basic
+
+/-!
+# `ne_zero` typeclass
+
+We create a typeclass `ne_zero n` which carries around the fact that `(n : R) ≠ 0`.
+
+## Main declarations
+
+* `ne_zero`: `n ≠ 0` as a typeclass.
+
+-/
+
+/-- A type-class version of `n ≠ 0`.  -/
+class ne_zero {R} [has_zero R] (n : R) : Prop := (out : n ≠ 0)
+
+lemma ne_zero.ne {R} [has_zero R] (n : R) [h : ne_zero n] : n ≠ 0 := h.out
+
+lemma ne_zero.ne' (n : ℕ) (R) [has_zero R] [has_one R] [has_add R] [h : ne_zero (n : R)] :
+  (n : R) ≠ 0 := h.out
+
+lemma ne_zero_iff {R : Type*} [has_zero R] {n : R} : ne_zero n ↔ n ≠ 0 :=
+⟨λ h, h.out, ne_zero.mk⟩
+
+lemma not_ne_zero {R : Type*} [has_zero R] {n : R} : ¬ ne_zero n ↔ n = 0 :=
+by simp [ne_zero_iff]
+
+namespace ne_zero
+
+variables {R M F : Type*} {r : R} {x y : M} {n p : ℕ} {a : ℕ+}
+
+instance pnat : ne_zero (a : ℕ) := ⟨a.ne_zero⟩
+instance succ : ne_zero (n + 1) := ⟨n.succ_ne_zero⟩
+
+lemma of_pos [preorder M] [has_zero M] (h : 0 < x) : ne_zero x := ⟨h.ne'⟩
+lemma of_gt  [canonically_ordered_add_monoid M] (h : x < y) : ne_zero y := of_pos $ pos_of_gt h
+
+instance char_zero [ne_zero n] [add_monoid M] [has_one M] [char_zero M] : ne_zero (n : M) :=
+⟨nat.cast_ne_zero.mpr $ ne_zero.ne n⟩
+
+lemma of_injective [semiring R] [ne_zero (n : R)] [semiring M] {f : R →+* M}
```

Comment:
You don't need semirings for this, right? Can you replace `f` by some term of a type `F` that carries `zero_hom_class`? And `R` and `M` should just be things that carry a `0`.

Reply:
I added this as another version. There's two issues related with this; firstly, `f (n : R)` is not defeq to `(n : M)`, so we can't just use these for casts without a separate version (although maybe we should switch the name for that instead of the priming I did); secondly, there's not currently enough generality in `data/nat/cast`, as it does not use `fun_like`. I will follow that up on another PR, and fix these and other results.